static/app/components/searchQueryBuilder/hooks/useQueryBuilderState.tsx

```diff
@@ -621,23 +628,104 @@ function updateFilterKey(
   };
 }
 
+const ALPHANUMERIC_REGEX = /[a-zA-Z0-9]/;
+
+export function replaceFreeTextTokens(
+  action: UpdateFreeTextAction | ReplaceTokensWithTextAction,
+  getFieldDefinition: FieldDefinitionGetter,
+  replaceRawSearchKeys: string[],
+  currentQuery: string
+) {
+  // if the free text is empty, return early
+  if (!action.text || action.text === '' || replaceRawSearchKeys.length === 0) {
+    return undefined;
+  }
+
+  // if the free text is not actually free text i.e. user entered a filter, return early
+  const actionTokens = parseQueryBuilderValue(action.text, getFieldDefinition) ?? [];
+  if (actionTokens.every(token => token.type !== Token.FREE_TEXT)) {
+    return undefined;
+  }
+
+  const tokens = parseQueryBuilderValue(currentQuery, getFieldDefinition) ?? [];
+  if (tokens.length === 0) {
+    return undefined;
+  }
+
+  // Single pass to find replace token
+  const primarySearchKey = replaceRawSearchKeys[0] ?? '';
+  // Find replace token and free text in single passes
+  let replaceToken: TokenResult<Token.FILTER> | undefined;
+  const freeTextToken = actionTokens.find(
+    token => token.type === Token.FREE_TEXT && ALPHANUMERIC_REGEX.test(token.value)
+  );
+
+  // Single pass through tokens for replace token
+  for (const token of tokens) {
+    if (token.type === Token.FILTER && token.text.includes(primarySearchKey)) {
+      replaceToken = token;
+      break; // Early exit once found
```

Comment:
nit: a few of these comments seem unnecessary, but one comment at the top of the function would be helpful to describe the intended behavior